server/filestore.go

```diff
@@ -3617,10 +3642,38 @@ func msgFromBuf(buf []byte, sm *StoreMsg, hh hash.Hash64) (*StoreMsg, error) {
 		sm.buf = append(sm.buf, data[slen:end]...)
 		sm.msg = sm.buf[0 : end-slen]
 	}
-	sm.subj, sm.seq, sm.ts = string(data[:slen]), seq, ts
+	sm.seq, sm.ts = seq, ts
+	// Treat subject a bit different to not reference underlying buf.
+	if slen > 0 {
+		sm.subj = mb.subjString(string(data[:slen]))
+	}
+
 	return sm, nil
 }
 
+// Lock should be held.
+func (mb *msgBlock) subjString(key string) string {
+	if len(key) == 0 {
+		return _EMPTY_
+	}
+
+	if lsubjs := len(mb.fs.cfg.Subjects); lsubjs > 0 {
+		if lsubjs == 1 && key == mb.fs.cfg.Subjects[0] {
+			return mb.fs.cfg.Subjects[0]
+		} else {
+			for _, subj := range mb.fs.cfg.Subjects {
+				if key == subj {
+					return subj
+				}
+			}
+		}
+	}
+	// Copy here to not reference underlying buffer.
```

Comment:
I don't understand this... subjString() is invoked with a string that is already a string (either from caller stack or made a string out of the underlying buffer), so I don't see how "key" is referencing the underlying buffer.